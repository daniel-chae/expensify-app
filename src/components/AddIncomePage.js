import React from 'react';
import IncomeForm from './IncomeForm';
import { connect } from 'react-redux';
import { startAddIncome } from '../actions/incomes';

export class AddIncomePage extends React.Component {
    onSubmit = (income) => {
        this.props.startAddIncome(income)
        this.props.history.push('/transaction'); //our component rendered thorugh react router comes with bunch of methods
      };
      render() {
        return (
          <div>
            <div className='page-header'>
              <div className = 'content-container'>
                <h1 className = 'page-header__title'>Add Income</h1>
              </div>
            </div>
            <div className = 'content-container'>
              <IncomeForm //What to do with added value is decided on parent component, we share the same expenseForm for adding and editing
              onSubmit={this.onSubmit}
              />       
            </div>
          </div>
        )
      }
    }

const mapDispatchToProps = (dispatch) => {
  return {
    startAddIncome: (income) => dispatch(startAddIncome(income))
  };
};
    
export default connect(undefined, mapDispatchToProps)(AddIncomePage);