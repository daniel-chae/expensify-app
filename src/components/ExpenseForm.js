import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { connect } from 'react-redux';

//Component that we render with React router comes with bunch of methods
export class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount /100).toString() : '',
            category: props.expense ? props.expense.category : 'Food',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            errorState: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(()=>({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        console.log(this.props)
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(()=>({ amount })); 
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt){
            this.setState(()=>({ createdAt }));
        }
    }
    onFocusChange=({focused})=>{
        this.setState(()=>({
            calendarFocused: focused
        }))
    }
    onCategoryChange = (e) => {
        const category = e.target.value;
        this.setState(()=>({ category }))
    }
    onSubmit = ((e) => {
        e.preventDefault(); //To prevent full page refresh
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                errorState: 'Please provide Description and amount'
            }))
        } else {
            this.setState(()=>({
                errorState: ''
            }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //to turn string into float, similarto parseInt
                createdAt: this.state.createdAt.valueOf(), //https://momentjs.com/docs/#/displaying/unix-timestamp-milliseconds/
                category: this.state.category,
                note: this.state.note
            });
        }
    })
    render() {
        return (
            <div>
                <form
                    className = "form"
                    onSubmit = {this.onSubmit}
                >
                    <p className = "form__error">{!!this.state.errorState && this.state.errorState}</p>
                    <input 
                        className='text-input'
                        type = 'text'
                        placeholder = 'Description'
                        autoFocus
                        value = {this.state.description}
                        onChange = {this.onDescriptionChange}
                    />
                    <input
                        className='text-input'
                        type = 'text'
                        placeholder = 'Amount'
                        value = {this.state.amount}
                        onChange = {this.onAmountChange}
                    />
                    <SingleDatePicker
                        date = {this.state.createdAt} //To display value from the state
                        onDateChange = {this.onDateChange} //To update value in the state
                        focused = {this.state.calendarFocused} //To control the date picker visibility.
                        onFocusChange = {this.onFocusChange} //When focus comes in and off this will be fired.
                        numberOfMonths = {1} // To control how many month to be shown in date picker
                        isOutsideRange = {() => false} // To allow the past date selection 
                    />
                    <select 
                        className = "select"
                        value = {this.state.category}
                        onChange = { this.onCategoryChange }
                    >
                        {this.props.categories.map((category)=>(<option key={category} value={category}>{category}</option>))}
                    </select>
                    <textarea
                        className='text-area'
                        placeholder='Add a note for your expense (optional)'
                        value = {this.state.note}
                        onChange = {this.onNoteChange}
                    >
                    </textarea>
                    <div>
                        <button className = 'button'>{this.props.expense? "Update Expense": "Add Expense"}</button>
                    </div>
                </form>
            </div>
        )
    }
}

export default ExpenseForm