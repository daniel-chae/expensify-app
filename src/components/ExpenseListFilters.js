import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'; //https://github.com/airbnb/react-dates

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onSortChange = (e) => {
        switch (e.target.value) {
            case 'date':
            return this.props.sortByDate()
            case 'amount':
            return this.props.sortByAmount()
        }
    };
    onTextChange = (e) => { 
        this.props.setTextFilter(e.target.value)
    };
    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused })
    };
    render() {
        return (
            <div className='content-container'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input 
                        className='text-input'
                        type='text' 
                        placeholder='Search Expenses'
                        value={ this.props.filters.text } //In class-based component this.props is used to access props
                        onChange={ this.onTextChange } //When value changes onChange is called with event argument
                        />
                    </div>
                    <div className='input-group__item'>
                        <select
                        className='select'
                        value={ this.props.filters.sortBy } 
                        onChange= { this.onSortChange }
                        >
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className='input-group__item'>
                        <DateRangePicker
                        startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                        endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                        onFocusChange={ this.onFocusChange } // PropTypes.func.isRequired,
                        numberOfMonths = {1}
                        isOutsideRange ={ ()=> false}
                        showClearDates = {true}
                        />
                    </div>
                </div>
            </div>
        )
    }

}
const mapStateToProps = (state) => ({
        filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => {dispatch(setTextFilter(text))},
    sortByDate: () => {dispatch(sortByDate())},
    sortByAmount: () => {dispatch(sortByAmount())},
    setStartDate: (startDate) => {dispatch(setStartDate(startDate))},
    setEndDate: (endDate) => {dispatch(setEndDate(endDate))}
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);