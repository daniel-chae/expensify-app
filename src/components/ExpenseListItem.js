import React from 'react';
import { NavLink } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div> 
        <NavLink to = {`/edit/${id}`}> 
            <h3>{description}</h3>
        </NavLink>
        <p>{amount} - {createdAt}</p>
    </div>
);

//Template string goes between two backtick `` variable can be part of string by using ${variable}

export default ExpenseListItem;