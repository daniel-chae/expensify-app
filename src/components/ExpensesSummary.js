import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expenses-total';
import numeral from 'numeral'; //http://numeraljs.com/

export const ExpensesSummary = (props) => {
    const filteredExpenses = selectExpenses(props.expenses, props.filters);
    const lengthFilteredExpenses = filteredExpenses.length
    const sumFilteredExpenses = getTotalExpenses(filteredExpenses);
    const formattedExpensesTotal = numeral(sumFilteredExpenses/100).format('$0,0.00')
    return (
        <div className="page-header">
            <div className="content-container">
                {lengthFilteredExpenses==0?(<h1 className="page-header__title">There is no expense.</h1>):
                    (<h1 className="page-header__title">
                    Viewing <span>{lengthFilteredExpenses}</span> expense totalling <span>{formattedExpensesTotal}</span>
                    </h1>)
                }
                    <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => ({
    expenses: state.expenses,
    filters: state.filters
})

export default connect(mapStateToProps)(ExpensesSummary)