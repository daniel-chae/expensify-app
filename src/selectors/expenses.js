import moment from 'moment'
// Get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true;//updated this to use moment
    const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day'): true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};
