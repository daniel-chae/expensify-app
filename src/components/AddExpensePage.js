import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense)
    this.props.history.push('/'); //our component rendered thorugh react router comes with bunch of methods
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm //What to do with added value is decided on parent component, we share the same expenseForm for adding and editing
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

//mapStateToProps, mapDispatchToProps
//in props.dispatch(addExpense(expense)) addExpense(expense) function is not
//passed as a prop. Thus, testing it with spy becomes trickier.
//If we use mapDispatchToProps, the function becomes the part of prop and
//normal spy can be used