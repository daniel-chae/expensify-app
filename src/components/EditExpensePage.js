import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (update) => {
    this.props.startEditExpense(this.props.expense.id, update)
    this.props.history.push('/')
  };
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  };
  render() {
    return (
      <div>
        <div className = 'page-header'>
          <div className = 'content-container'>
            <h1 className = 'page-header__title'>Edit Expense</h1>
          </div>
        </div>
        <div className = "content-container">
          <ExpenseForm 
          expense = {this.props.expense}
          categories = {this.props.categories}
          onSubmit = {this.onSubmit}
          />       
          <button className = 'button button--bottom button--secondary' onClick = {this.onClick}>Remove Expense</button>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense)=>{
      return expense.id === props.match.params.id;
    }),
    categories: state.settings.categories
  }
}

const mapDispatchToProp = (dispatch) => {
  return {
    startEditExpense: (id, update) => dispatch(startEditExpense(id, update)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))    
  }
};

export default connect(mapStateToProps, mapDispatchToProp)(EditExpensePage);
