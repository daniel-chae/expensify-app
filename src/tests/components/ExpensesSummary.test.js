import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import { filters, altFilters, testFilters } from '../fixtures/filters';

test('should match the snapshot for ExpensesSummary with filter variation 1', () => {
    const wrapper = shallow(<ExpensesSummary expenses = {expenses} filters = {filters}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should match the snapshot for ExpensesSummary with filter variation 2', () => {
    const wrapper = shallow(<ExpensesSummary expenses = {expenses} filters = {testFilters}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should match the snapshot for ExpensesSummary with filter variation 3', () => {
    const wrapper = shallow(<ExpensesSummary expenses = {expenses} filters = {altFilters}/>);
    expect(wrapper).toMatchSnapshot();
});