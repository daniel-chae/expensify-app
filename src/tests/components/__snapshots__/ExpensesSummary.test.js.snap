// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for ExpensesSummary with filter variation 1 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       expense totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should match the snapshot for ExpensesSummary with filter variation 2 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense totalling 
      <span>
        $1,095.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should match the snapshot for ExpensesSummary with filter variation 3 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      There is no expense.
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
